checker-rs/src/notebook_client.rs

```diff
@@ -0,0 +1,351 @@
+        net::TcpStream,
+    },
+    CheckerError, CheckerRequest, CheckerResult,
+};
+use std::io;
+
+use tracing::{debug, error, info, instrument, warn};
+/// reads until the delimiter given in delim is found or EOF if found or another Error occurs
+///
+/// # Return
+/// Number of bytes read
+pub async fn read_until_slice<'a, T: AsyncBufRead + Unpin>(
+    stream: &'a mut T,
+    delim: &'a [u8],
+    buf: &'a mut Vec<u8>,
+) -> io::Result<usize> {
+    let mut bytes_read: usize = 0;
+    let mut inc_read;
+
+    let delim_end = *delim.last().expect("DELIMITER must not be empty");
+
+    loop {
+        inc_read = stream.read_until(delim_end, buf).await?;
+        if inc_read == 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "Delimiter was not found",
+            ));
+        }
+        bytes_read += inc_read;
+        if bytes_read >= delim.len() && &buf[(buf.len() - delim.len())..buf.len()] == delim {
+            return Ok(bytes_read);
+        }
+    }
+}
+
+enum CheckerErrorUnfilled<T: std::fmt::Debug + 'static> {
+    Mumble,
+    Offline,
+    InternalError(T),
+}
+
+impl From<io::Error> for CheckerErrorUnfilled<io::Error> {
+    fn from(e: std::io::Error) -> CheckerErrorUnfilled<std::io::Error> {
+        match e.kind() {
+            io::ErrorKind::ConnectionRefused
+            | io::ErrorKind::ConnectionAborted
+            | io::ErrorKind::ConnectionReset => CheckerErrorUnfilled::Offline,
+            io::ErrorKind::UnexpectedEof => CheckerErrorUnfilled::Mumble,
+            _ => CheckerErrorUnfilled::InternalError(e),
+        }
+    }
+}
+
+impl<T: std::fmt::Debug + 'static> CheckerErrorUnfilled<T> {
+    fn with_message(self, err_msg: &'static str) -> CheckerError {
+        match self {
+            CheckerErrorUnfilled::Offline => CheckerError::Offline(err_msg),
+            CheckerErrorUnfilled::Mumble => CheckerError::Mumble(err_msg),
+            CheckerErrorUnfilled::InternalError(err) => {
+                error!("Internal error: {:?}", err);
+                CheckerError::InternalError("Internal Error")
+            }
+        }
+    }
+}
+
+pub fn bytes_debug_repr(bytes: &[u8]) -> String {
+    let mut bytes_repr = "b'".to_owned();
+    for b in bytes {
+        let b_escaped: Vec<_> = std::ascii::escape_default(*b).collect();
+        bytes_repr.push_str(std::str::from_utf8(&b_escaped).unwrap());
+    }
+    bytes_repr.push('\'');
+    bytes_repr
+}
+
+#[derive(Debug)]
+pub struct NotebookClient {
+    pub conn: BufStream<TcpStream>,
+    pub user: Option<NotebookUser>,
+}
+
+const PROMPT: &[u8] = b"\n> ";
+
+impl NotebookClient {
+    const SERVICE_PORT: u16 = 2323;
+
+    #[instrument("CONNECT")]
+    pub async fn connect(request: &CheckerRequest) -> CheckerResult<Self> {
+        info!("Connecting to service");
+        let mut conn =
+            match TcpStream::connect(format!("{}:{}", request.address, Self::SERVICE_PORT)).await {
```

Comment:
AFAIK takes `A` that can also take an Iterator/Tuple directly, instead of calling format.